cpp/ql/src/semmle/code/cpp/commons/CommonType.qll

```diff
@@ -63,26 +63,264 @@ class Ptrdiff_t extends Type {
   override string getAPrimaryQlClass() { result = "Ptrdiff_t" }
 }
 
+/**
+ * A common base type for describing the C/C++ fixed-width numeric types
+ */
+abstract class FixedWidthIntegralType extends UserType {
```

Comment:
We've made a big push to avoid public abstract classes in the base libraries, including this file. See https://github.com/github/codeql-c-team/issues/60 for some motivation and context. Abstract classes are great for the library-internal implementation but painful and dangerous to work with for consumers of the library. For code in the base libraries, the balance is toward reducing surprises for the library's consumers.

Reply:
What approach do you recommend here?  This class won't behave correctly in its current form without `abstract`.

Reply:
I'd suggest going with a union approach similar to what we've done here: https://github.com/github/codeql/pull/3940

i.e., make a `TFixedWidthIntegralType` that's a union of all the classes that currently implement the abstract class, and then make a `FixedWidthIntegralType` class that extends `TFixedWidthIntegralType`.

Reply:
@jbj @geoffw0 @MathiasVP -- Please see the latest push to this PR which incorporates a bunch of suggestions from @jbj for resolving this issue. Please let me know so we can come to a consensus. 

Also, I'm failing the changelog check but it doesn't look like it's been updated in a while. Shall I create a new directory and note these changes (if so, what directory/version to add?) or omit the changelog entry?

Reply:
reading @MathiasVP -- I'll update and push again.

Reply:
We've _changed_ (🥁) the way we do change-notes (which is why that directory hasn't been updated in a while). See an example [here](https://github.com/github/codeql/pull/5859) on how we do it now (specifically, the file `cpp/change-notes/2021-10-05-comparison-with-wider-type.md ` created by that PR).

Reply:
Thanks @MathiasVP!